beamer/check/commands.py

```diff
@@ -264,3 +264,166 @@ def initiate_l1_invalidations(
+
+    abi_manager = ABIManager(abi_dir)
+    rpc_info = beamer.util.load_rpc_info(rpc_file)
+
+    with file.open("rt") as f:
+        data = json.load(f)
+    invalidations = apischema.deserialize(list[Invalidation], data)
+
+    # Pairs (proof_source, proof_target) for which an invalidation has already been
+    # verified.
+    verified = set()
+
+    errors = 0
+    chain_pairs = set()
+    for idx, invalidation in enumerate(invalidations):
+        pair = invalidation.proof_source, invalidation.proof_target
+        chain_pairs.add(pair)
+
+        if pair in verified:
+            log.info(
+                "Verified invalidation already exists for chain pair, skipping",
+                idx=idx,
+                proof_source=pair[0],
+                proof_target=pair[1],
+            )
+            continue
+
+        # First check the finalization timestamp.
+        if (
+            invalidation.finalization_timestamp is None
+            or datetime.utcnow() <= invalidation.finalization_timestamp
+        ):
+            log.error(
+                "Invalidation not yet finalized, skipping",
+                idx=idx,
+                proof_source=pair[0],
+                proof_target=pair[1],
+            )
+            errors += 1
+            continue
+
+        deployment = beamer.artifacts.load(artifacts_dir, invalidation.proof_target)
+        url = rpc_info[invalidation.proof_target]
+        w3 = make_web3(url, account)
+        assert w3.eth.chain_id == invalidation.proof_target
+        log.info("Connected to RPC", url=url)
+
+        # Check the request manager directly to see if it has been already verifed.
+        request_manager = obtain_contract(w3, abi_manager, deployment, "RequestManager")
+        if request_manager.functions.isInvalidFill(
+            invalidation.request_id, invalidation.fill_id
+        ).call():
+            log.info(
+                "Invalidation verified successfully",
+                idx=idx,
+                proof_source=pair[0],
+                proof_target=pair[1],
+            )
+            verified.add(pair)
+            continue
+
+        log.debug("Starting relayer", txhash=invalidation.txhash)
+        try:
+            run_relayer_for_tx(
+                l1_rpc=rpc_info[deployment.base.chain_id],
+                l2_relay_from_rpc_url=rpc_info[invalidation.proof_source],
+                l2_relay_to_rpc_url=rpc_info[invalidation.proof_target],
+                account=account,
+                tx_hash=invalidation.txhash,
+            )
+        except RelayerError as exc:
+            log.error(
+                "Relayer failed", idx=idx, exc=exc, proof_source=pair[0], proof_target=pair[1]
+            )
+            errors += 1
+        else:
+            log.info("Relayer succeeded", thxash=invalidation.txhash)
+
+            # Check the request manager, the invalidation should have arrived.
+            if request_manager.functions.isInvalidFill(
```

Comment:
This will only work after the relayer is checking the request manager, right?

Reply:
Yes, that assumes the relayer change is done, as discussed with Gabriel.